libcob/fileio.c

```diff
@@ -1686,29 +1792,76 @@ cob_chk_file_mapping (cob_file *f, char *filename)
 	}
 	/* First element completed, loop through remaining */
 	/* elements delimited by slash */
-	/* Check each for $ mapping */
+	/* Check only for $ from now on; includes the DD_xx/dd_xx/xx mapping */
+	src = NULL;
 	for (; ;) {
 		p = strtok (orig, "/\\");
 		if (!p) {
 			break;
 		}
 		if (!orig) {
-			if (dollar) {
-				dollar = 0;
-			} else {
+			if (!dollar) {
 				strcat (file_open_buff, SLASH_STR);
 			}
 		} else {
 			orig = NULL;
 		}
-		if (*p == '$' && (src = cob_chk_file_env (f, p + 1)) != NULL) {
-			strncat (file_open_buff, src, (size_t)COB_FILE_MAX);
+		if (*p != '$') {
+			dollar = 0;
 		} else {
+			dollar = 1;
+			p++;
+		}
+		if (dollar && (src = cob_chk_file_env (f, p)) != NULL) {
+			strncat (file_open_buff, src, (size_t)COB_FILE_MAX);
+			src = NULL;
+		} else if (!dollar) {
 			strncat (file_open_buff, p, (size_t)COB_FILE_MAX);
+			src = NULL;
+		} else {
+			src = p - 1;
 		}
 	}
+	/* if we have a final $something that cannot be resolved - use as plain name */
+	if (src) {
+		strncat (file_open_buff, src, (size_t)COB_FILE_MAX);
+	}
 	strcpy (file_open_name, file_open_buff);
 	cob_free (saveptr);
+
+	if (looks_absolute (file_open_name)) {
+		return;
+	}
+	/* apply COB_FILE_PATH if set (similar to ACUCOBOL's FILE-PREFIX) */
+	if (file_paths) {
+		for(k=0; file_paths[k] != NULL; k++) {
+			snprintf (file_open_buff, (size_t)COB_FILE_MAX, "%s%c%s",
+				  file_paths[k], SLASH_CHAR, file_open_name);
+			file_open_buff[COB_FILE_MAX] = 0;
+			if (access (file_open_buff, F_OK) == 0) {
+				break;
+			}
+#if defined(WITH_CISAM) || defined(WITH_DISAM) || defined(WITH_VBISAM) || defined(WITH_VISAM)
+			/* ISAM may append '.dat' to file name */
+			snprintf (file_open_buff, (size_t)COB_FILE_MAX, "%s%c%s.dat",
+				  file_paths[k], SLASH_CHAR, file_open_name);
+			file_open_buff[COB_FILE_MAX] = 0;
+			if (access (file_open_buff, F_OK) == 0) {
+				snprintf (file_open_buff, (size_t)COB_FILE_MAX, "%s%c%s",
+					  file_paths[k], SLASH_CHAR, file_open_name);
+				file_open_buff[COB_FILE_MAX] = 0;
+				break;
+			}
+#endif
+		}
+		if (file_paths[k] == NULL) {
+			snprintf (file_open_buff, (size_t)COB_FILE_MAX, "%s%c%s",
+				  file_paths[0], SLASH_CHAR, file_open_name);
+			file_open_buff[COB_FILE_MAX] = 0;
+		}
+		strncpy (file_open_name, file_open_buff, (size_t)COB_FILE_MAX);
+	}
```

Comment:
Do we need that twice? Otherwise this duplicated "nearly identical" logic (file-path and .dat) should really be moved out to a function called in both places - please refactor that for the merge commit.

Reply:
Yeah, the "simple" and "complex" cases share some common code.
Before rev 3944, GC3 applied COB_FILE_PATH only to the simple case (and so did GC4 HEAD).
Rev 3944 made GC3 also apply COB_FILE_PATH to the complex case, using similar (small) code in both cases.
GC4 code to apply COB_FILE_PATH being larger, merging 3944 in GC4 resulted in this code increase / duplication.
I'll refactor appropriately.

Reply:
I remember why I haven't refactored that straight away : in case subsequent commits modify the same code, conflicts will be much easier to handle. I was thinking about keeping the refactoring for after all the patches are merged...

Reply:
Change done.